### DISCUSSION TITLE
How to set time-out when client.TransactionRequest?

### DISCUSSION THREAD

**Discussion by zcc0077** [emiago/sipgo#82] [Q&A] [answered]
I found there may be a default timeout setting 
```
[2024-03-29 14:06:08.802] INF Starting transaction req=OPTIONS
[2024-03-29 14:06:40.806] ERR Fail to get response error="transaction died"
```
I want to set it manually, didn't find out how to achieve this. I tried 
```
ctx, cancel := context.WithTimeout(context.Background(), time.Duration(5)*time.Second)
defer cancel()
```
before `client.TransactionRequest(ctx, req)`, seems no effect.
And I tried 
```
func getResponse(tx sip.ClientTransaction) (*sip.Response, error) {
    select {
    case <-time.After(time.Duration(5) * time.Second):
        return nil, fmt.Errorf("time out exceed")
    case <-tx.Done():
        return nil, fmt.Errorf("transaction died")
    case res := <-tx.Responses():
        return res, nil
    }
}
```
Still no effect.
Could you give me some hints? Thanks a lot :)

**Comment by zcc0077** [comment]
I found this is interesting, here is my scenario, sometimes I need to run multi-times in one program lifecycle to send udp,tcp and tls OPTION.
when I use `ctx, cancel := context.WithTimeout(context.Background(), time.Duration(5)*time.Second)`
to test tcp and tls, seems timeout can work, but when it comes to udp, timeout takes no effect.
```
[2024-03-29 14:36:23.497] INF Protocol: udp, timeout: 5
[2024-03-29 14:36:23.499] INF OPTIONS sip:xx.xx.xx.xx:5060 SIP/2.0
[2024-03-29 14:36:23.500] INF Starting transaction req=OPTIONS
[2024-03-29 14:36:55.501] ERR Fail to get response error="transaction died"
[2024-03-29 14:41:09.943] INF Protocol: tcp, timeout: 5
[2024-03-29 14:41:09.943] INF OPTIONS sip:xx.xx.xx.xx:5060 SIP/2.0
[2024-03-29 14:41:10.157] INF Starting transaction req=OPTIONS
^C[2024-03-29 14:41:36.701] WRN Received signal: interrupt
```
and when this happened, tcp and tls all lose timeout limitation...

**Reply by emiago** [reply, marked as answer]
hi @zcc0077 . I am not clear where you want to timeout. 
`TransactionRequest` `ctx` has only effect on building connection, but not on response. It returns you transaction object on which you wait for response until transaction timeouts from SIP timers. More on this below
Ofcourse you can wrap this with timeout on how long you want to wait. 

sipgo is currently more on low level API, comparing like std http, I avoid wrapping too much. I see some potential like wrapping this stuff for things where you want to just send request and get response. In this case ctx behavior would be extended.


As for transaction timeouts. You need to watch also on SIP timers. So transaction can timeout `tx.Done()` due to SIP timers, and waiting for response should be stopped in this case. Therefore you always must watch on this `tx.Done()` 

Changing SIP timers is also possible

**Reply by zcc0077** [reply]
thanks for the quick response :) I'll take a shot.

**Reply by zcc0077** [reply]
```
	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(data.GetConfiguration().SipTimeOut)*time.Second)
	sip.Timer_B = time.Duration(data.GetConfiguration().SipTimeOut) * time.Second
	defer cancel()
```
I use these two methods to limit timeout, only one is not enough lol.

**Reply by emiago** [reply]
SIP timers should be your global thing. So doing like this it will affect all transactions, and they onlz should be changed on some bootstrap. 

So generrally there is also nothing wrong doing this 
```go
cxt  := context.WithTimeout(...) 
tx, err := client.TransactionRequest(ctx, req)  // ctx is here for connections. For UDP has no big impact as it is just creating listener. For TCP it is TCP handshake. For TLS it is TCP + TLS handshake

defer tx.Terminate()  // Always close transaction
select {
   case <-ctx.Done(): 
             // my timeout. Transaction retries are too long 
   case <-tx.Done():
         // Should not be that quick It has error or my SIP timer is fast
         err := tx.Err()
    case res:= -tx.Response():
          // handle response
}            
```

So you have control to terminate transactation when ever you want

**Reply by emiago** [reply]
The missing part is probably. WriteDeadline when passing request transmission, so if you are having issues with transmission of packets this could block

**Reply by zcc0077** [reply]
sry I can't get you , which missing part? what is WriteDeadline?  Currently this code is working for me just to test the transmission of OPTION with customized timeout.

**Reply by emiago** [reply]
library is currently not taking ctx timeout when passing packets (syscall) . WriteDeadline is not set. In practice it is sign a much bigger problem, but for simple use you do not need this.
Anyway this will be fixed in some release
